File: test/test_speaker_encoder_e2e.py
"""Layer-2 E2E cosine gate for the speaker encoder.

See ``docs/spec/speaker-encoder-contract.md``. This test is **opt-in** and
skips by default — it activates only when:

1. The fixture ``test/fixtures/speaker_encoder_golden.json`` contains an
   ``e2e_cosine_gate`` block (i.e. someone ran the generator with
   ``--encoder-onnx`` populated), AND
2. A local encoder ONNX is available — either via env var
   ``PIPER_SPEAKER_ENCODER_ONNX_PATH`` or via HF Hub download (the latter
   only when ``PIPER_SPEAKER_ENCODER_E2E=1`` is set, to avoid hitting the
   network on every CI run).

Test semantics: compute embedding from the reference WAV using the local
encoder, assert ``cosine(actual, expected) >= cosine_threshold``. The
threshold (0.999) is permissive on purpose — see the spec for the rationale
on why byte-equality across ORT execution providers is not feasible.
"""

from __future__ import annotations

import json
import os
from pathlib import Path

import numpy as np
import pytest

FIXTURE_PATH = Path(__file__).parent / "fixtures" / "speaker_encoder_golden.json"


def _load_fixture() -> dict:
    with open(FIXTURE_PATH, encoding="utf-8") as f:
        return json.load(f)


def _resolve_encoder_path(gate: dict) -> Path | None:
    """Locate the encoder ONNX. Returns None to signal `skip`.

    Priority:
      1. ``PIPER_SPEAKER_ENCODER_ONNX_PATH`` env var (explicit override).
      2. HF Hub download, only when ``PIPER_SPEAKER_ENCODER_E2E=1``.
    """
    env_path = os.environ.get("PIPER_SPEAKER_ENCODER_ONNX_PATH")
    if env_path:
        p = Path(env_path)
        if p.exists():
            return p
        pytest.skip(f"PIPER_SPEAKER_ENCODER_ONNX_PATH={env_path} does not exist")

    if os.environ.get("PIPER_SPEAKER_ENCODER_E2E") != "1":
        return None  # opt-in only — skip silently

    try:
        from huggingface_hub import hf_hub_download  # noqa: PLC0415
    except ImportError:
        pytest.skip("huggingface_hub not installed; cannot download encoder ONNX")

    enc = gate["encoder_onnx"]
    return Path(
        hf_hub_download(
            repo_id=enc["hf_repo"],
            filename=enc["hf_filename"],
            revision=enc["hf_revision"],
        )
    )


def _verify_sha256(path: Path, expected: str) -> None:
    import hashlib  # noqa: PLC0415

    h = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    actual = h.hexdigest()
    if actual != expected:
        pytest.fail(
            f"encoder ONNX sha256 mismatch (silent upstream replacement?):\n"
            f"  expected: {expected}\n"
            f"  actual:   {actual}\n"
            f"  path:     {path}"
        )


def _cosine(a: np.ndarray, b: np.ndarray) -> float:
    a = np.asarray(a, dtype=np.float32)
    b = np.asarray(b, dtype=np.float32)
    na = float(np.linalg.norm(a))
    nb = float(np.linalg.norm(b))
    if na == 0 or nb == 0:
        return 0.0
    return float(np.dot(a, b) / (na * nb))


def test_e2e_cosine_gate_against_pinned_embedding() -> None:
    fixture = _load_fixture()
    gate = fixture.get("e2e_cosine_gate")
    if gate is None:
        pytest.skip(
            "fixture has no e2e_cosine_gate block — generator was run without "
            "--encoder-onnx; the layer-1 mel parity tests still apply"
        )

    encoder_path = _resolve_encoder_path(gate)
    if encoder_path is None:
        pytest.skip(
            "encoder ONNX not available locally and PIPER_SPEAKER_ENCODER_E2E "
            "is not set — opt-in test, skipping by default"
        )

    expected_sha = gate["encoder_onnx"].get("sha256")
    if expected_sha and expected_sha != "":
        _verify_sha256(encoder_path, expected_sha)

    # Lazy imports — ``onnxruntime`` is only required on the active path.
    try:
        import onnxruntime  # noqa: F401, PLC0415
    except ImportError:
        pytest.skip("onnxruntime not installed; cannot exercise E2E gate")

    # Re-use the generator's compute helper so the test stays a thin shell.
    from generate_speaker_encoder_golden import _compute_e2e_embedding  # noqa: PLC0415

    wav_path = Path(gate["reference_wav"]["path"])
    if not wav_path.is_absolute():
        wav_path = (Path(__file__).parent.parent / wav_path).resolve()
    if not wav_path.exists():
        pytest.skip(f"reference WAV not found at {wav_path}")

    actual_embedding, _ = _compute_e2e_embedding(encoder_path, wav_path)
    expected_embedding = np.asarray(
        gate["expected_embedding"]["values"], dtype=np.float32
    )

    assert actual_embedding.shape == expected_embedding.shape, (
        f"embedding dim drift: actual={actual_embedding.shape}, "
        f"expected={expected_embedding.shape}"
    )

    cos = _cosine(actual_embedding, expected_embedding)
    threshold = float(gate["cosine_threshold"])
    assert cos >= threshold, (
        f"cosine gate failed: cos={cos:.6f} < threshold={threshold:.6f}\n"
        f"  encoder: {encoder_path}\n"
        f"  WAV:     {wav_path}\n"
        f"  fixture sha256={gate['encoder_onnx'].get('sha256', '<unset>')}"
    )

File: src/rust/piper-core/tests/test_speaker_encoder_e2e.rs
//! Layer-2 E2E cosine gate for the speaker encoder.
//!
//! Mirrors `test/test_speaker_encoder_e2e.py`. See
//! `docs/spec/speaker-encoder-contract.md` for the contract.
//!
//! This test is **opt-in** and skips by default — it activates only when:
//! 1. The fixture `test/fixtures/speaker_encoder_golden.json` contains an
//!    `e2e_cosine_gate` block, AND
//! 2. `PIPER_SPEAKER_ENCODER_ONNX_PATH` env var points at a local encoder
//!    ONNX (we do **not** auto-download from HF in Rust tests — opt-in only).
//!
//! Run: `PIPER_SPEAKER_ENCODER_ONNX_PATH=/path/to/encoder.onnx \
//!       cargo test --features onnx --test test_speaker_encoder_e2e`
//!
//! The test body is gated behind the `onnx` feature because `SpeakerEncoder`
//! itself is. With `--no-default-features` or without `onnx`, the test
//! becomes a no-op (still compiles, still passes — just skipped).

#![cfg(feature = "onnx")]

use std::fs;
use std::path::{Path, PathBuf};

use serde::Deserialize;
use sha2::{Digest, Sha256};

use piper_plus::speaker_encoder::SpeakerEncoder;

#[derive(Deserialize)]
struct Fixture {
    e2e_cosine_gate: Option<E2EBlock>,
}

#[derive(Deserialize)]
struct E2EBlock {
    encoder_onnx: EncoderRef,
    reference_wav: ReferenceWav,
    expected_embedding: ExpectedEmbedding,
    cosine_threshold: f32,
}

#[derive(Deserialize)]
struct EncoderRef {
    sha256: Option<String>,
    #[allow(dead_code)]
    hf_repo: String,
    #[allow(dead_code)]
    hf_revision: String,
    #[allow(dead_code)]
    hf_filename: String,
}

#[derive(Deserialize)]
struct ReferenceWav {
    path: String,
    #[allow(dead_code)]
    sha256: Option<String>,
}

#[derive(Deserialize)]
struct ExpectedEmbedding {
    #[allow(dead_code)]
    dim: usize,
    values: Vec<f32>,
    #[allow(dead_code)]
    checksum: Option<String>,
}

fn fixture_path() -> PathBuf {
    let crate_root = env!("CARGO_MANIFEST_DIR");
    PathBuf::from(crate_root)
        .join("..")
        .join("..")
        .join("..")
        .join("test")
        .join("fixtures")
        .join("speaker_encoder_golden.json")
}

fn repo_root() -> PathBuf {
    let crate_root = env!("CARGO_MANIFEST_DIR");
    PathBuf::from(crate_root).join("..").join("..").join("..")
}

fn sha256_file(path: &Path) -> std::io::Result<String> {
    let bytes = fs::read(path)?;
    let mut h = Sha256::new();
    h.update(&bytes);
    Ok(format!("{:x}", h.finalize()))
}

fn cosine(a: &[f32], b: &[f32]) -> f32 {
    let dot: f32 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    let na: f32 = a.iter().map(|x| x * x).sum::<f32>().sqrt();
    let nb: f32 = b.iter().map(|x| x * x).sum::<f32>().sqrt();
    if na == 0.0 || nb == 0.0 {
        0.0
    } else {
        dot / (na * nb)
    }
}

#[test]
fn e2e_cosine_gate_against_pinned_embedding() {
    let raw = match fs::read_to_string(fixture_path()) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("[skip] cannot read fixture: {e}");
            return;
        }
    };
    let fixture: Fixture = match serde_json::from_str(&raw) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("[skip] cannot parse fixture: {e}");
            return;
        }
    };
    let gate = match fixture.e2e_cosine_gate {
        Some(g) => g,
        None => {
            eprintln!(
                "[skip] fixture has no e2e_cosine_gate block — generator was \
                 run without --encoder-onnx; layer-1 mel parity tests still apply"
            );
            return;
        }
    };

    let encoder_path = match std::env::var("PIPER_SPEAKER_ENCODER_ONNX_PATH") {
        Ok(p) => PathBuf::from(p),
        Err(_) => {
            eprintln!(
                "[skip] PIPER_SPEAKER_ENCODER_ONNX_PATH not set — opt-in test, \
                 skipping by default"
            );
            return;
        }
    };
    if !encoder_path.exists() {
        panic!(
            "PIPER_SPEAKER_ENCODER_ONNX_PATH={} does not exist",
            encoder_path.display()
        );
    }

    if let Some(expected_sha) = gate.encoder_onnx.sha256.as_deref()
        && !expected_sha.is_empty()
    {
        let actual_sha = sha256_file(&encoder_path).expect("sha256");
        assert_eq!(
            actual_sha, expected_sha,
            "encoder ONNX sha256 mismatch (silent upstream replacement?)"
        );
    }

    let wav_path = {
        let p = PathBuf::from(&gate.reference_wav.path);
        if p.is_absolute() {
            p
        } else {
            repo_root().join(p)
        }
    };
    if !wav_path.exists() {
        eprintln!("[skip] reference WAV not found at {}", wav_path.display());
        return;
    }

    let mut encoder = SpeakerEncoder::new(&encoder_path).expect("load encoder");
    let actual = encoder.encode_file(&wav_path).expect("encode");

    assert_eq!(
        actual.len(),
        gate.expected_embedding.values.len(),
        "embedding dim drift"
    );

    let cos = cosine(&actual, &gate.expected_embedding.values);
    assert!(
        cos >= gate.cosine_threshold,
        "cosine gate failed: cos={cos:.6} < threshold={:.6} \
         (encoder={}, wav={})",
        gate.cosine_threshold,
        encoder_path.display(),
        wav_path.display(),
    );
}
